File: neetcode/blind75/arrays_and_hashing/longest_consecutive_sequence.py
# https://neetcode.io/problems/longest-consecutive-sequence

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        seq_dict = dict()

        # sort ascending
        nums.sort()
        max_seq = 0

        # for each int, check if the previous int exists in the seq_dict; if yes, add to dict with +1 else 1.
        for n in nums:
            seq_val = seq_dict.get(n-1, 0) + 1
            seq_dict[n] = seq_val
            max_seq = max(max_seq, seq_val)

        #print (seq_dict)
        # return max value
        return max_seq

File: neetcode/neet150_JS/Arrays-and-Hashing/longest-consecutive-sequence.js
// https://neetcode.io/problems/longest-consecutive-sequence

class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    longestConsecutive(nums) {
        if (nums.length === 0) {
            return 0
        }

        let max = 0
        const streaks = new Map()

        for (let i = 0; i < nums.length; i ++) {
            if (!streaks.has(nums[i])) {
                streaks.set(nums[i], (streaks.get(nums[i] - 1) || 0) + (streaks.get(nums[i] + 1) || 0) + 1)

                streaks.set(nums[i] - (streaks.get(nums[i] - 1) || 0), streaks.get(nums[i]))
                streaks.set(nums[i] + (streaks.get(nums[i] + 1) || 0), streaks.get(nums[i]))
            
                max = Math.max(max, streaks.get(nums[i]))
            }
        }

        return max
    }
}
